src/package.cc

```diff
@@ -60,6 +60,99 @@ Package::Package(std::string name, double fill_min, double fill_max,
         throw ValueError("can't create a new package with name 'unpackaged'");
       }
   }
+    if (strategy != "first" && strategy != "equal") {
+      throw ValueError("Invalid strategy for package: " + strategy_);
+    }
+}
+
+TransportUnit::Ptr TransportUnit::unrestricted_ = NULL;
+
+TransportUnit::Ptr TransportUnit::Create(std::string name, int fill_min, int fill_max, std::string strategy) {
+  if (fill_min < 0 || fill_max < 0) {
+    throw ValueError("fill_min and fill_max must be non-negative");
+  }
+  else if (fill_min > fill_max) {
+    throw ValueError("fill_min must be less than or equal to fill_max");
+  }
+  Ptr t(new TransportUnit(name, fill_min, fill_max, strategy));
+  return t;
+}
+
+// singleton pattern: 
+// if the static member is not yet set, create a new object
+// otherwise return the object that already exists
+TransportUnit::Ptr& TransportUnit::unrestricted() {
+  if (!unrestricted_) {
+    unrestricted_ = Ptr(new TransportUnit(unrestricted_name_));
+  }
+  return unrestricted_;
+}
+
+int TransportUnit::GetTransportUnitFill(int qty) {
+  if (qty < fill_min_) {
+    // less than one TransportUnit of material available
+    return 0;
+  }
+
+  if (strategy_ == "first") {
+    return fill_max_;
+  } else if (strategy_ == "equal" || strategy_ == "hybrid") {
+    // int division automatically rounds down. don't need floor in min, and 
+    // get ceil by hand instead 
+    int num_at_min_fill = qty / fill_min_;
+    int num_at_max_fill = (qty + fill_max_ - 1) / fill_max_;
+
+    if (num_at_min_fill >= num_at_max_fill) {
+      // all material *might* fit transport units. However, this is more
+      // challenging than packages because transport units are discrete. 
+      double dbl_fill_mass = (double)qty / (double)num_at_max_fill;
+      return std::floor(dbl_fill_mass);
+    }
+    // some material will remain unrestricted, fill up as many transport
+    // units as possible. Or, perfect fill is possible but not with integer
+    // fill (see above). 
+  }
+  return fill_max_;
+}
+
+int TransportUnit::MaxShippablePackages(int pkgs) {
+  int TU_fill;
+  int shippable = 0;
+
+  if (pkgs == 0 || pkgs < fill_min_) {
+    return 0;
+
+  } else if (name_ == unrestricted_name_) {
+    return pkgs;
+
+  } else if (strategy_ == "first" || strategy_ == "equal") {
+    TU_fill = GetTransportUnitFill(pkgs);
+    if (TU_fill == 0) {
+      return 0;
+    }
+    shippable = std::min(pkgs, (pkgs / TU_fill) * TU_fill);
```

Comment:
How is it possible for `(pkgs / TU_fill) * TU_fill` to be greater than `pkgs`?

Reply:
It gets a bit funky when package max fill is at the numeric limit